sql: stop using txn.OrigTimestamp to reset TableCollection

vivekmenezes opened the pull request:
Release note: None

#23718

vivekmenezes committed:
sql: Stop using txn.OrigTimestamp to reset TableCollection

We use the transaction OrigTimestamp to figure out if a
transaction is being retried, triggering a reset of
TableCollection. This logic would get triggered when
running PREPARE within a transaction, because PREPARE
uses it's own transaction and thus has a different
OrigTimestamp.

Removed this logic because transaction retries trigger
resetting the TableCollection in the new conn-executor code.

fixes #24578

Release note sql: fix PREPARE hanging when run in the same transaction
as a CREATE TABLE

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/23816)
<!-- Reviewable:end -->

vivekmenezes commented:
@andreimatei I'd like to add some testing for this. Is there a test that tests transaction retries and the fallout from that?

andreimatei commented:
there's `txn_restart_test.go` and `txn_retry` in logictests. You might make use of `crdb_internal.force_retry()`.

---

Review status: 0 of 2 files reviewed at latest revision, all discussions resolved, some commit checks failed.

---

*[pkg/sql/logictest/testdata/logic_test/txn, line 14 at r1](https://beta.reviewable.io/reviews/cockroachdb/cockroach/23816#-L9lG6ze7n1pgtr8QDD0:-L9lG6ze7n1pgtr8QDD1:b-o8shoh) ([raw file](https://github.com/cockroachdb/cockroach/blob/0fe903cf72156a3074c1b47aaa40ee1273ca7ce7/pkg/sql/logictest/testdata/logic_test/txn#L14)):*
> ```
> 
> query TT
> SELECT * FROM kv
> ```

what's going on here?

---

*[pkg/sql/logictest/testdata/logic_test/txn, line 49 at r1](https://beta.reviewable.io/reviews/cockroachdb/cockroach/23816#-L9lG_Q-2Fc5hVMw6uQY:-L9lG_Q-2Fc5hVMw6uQZ:boy8whw) ([raw file](https://github.com/cockroachdb/cockroach/blob/0fe903cf72156a3074c1b47aaa40ee1273ca7ce7/pkg/sql/logictest/testdata/logic_test/txn#L49)):*
> ```
> a c
> 
> statement ok
> ```

Please explain what we're testing. Also I'd move this to the prepare tests. Also consider using a subtest (we support those in logic tests now).

---

*[pkg/sql/logictest/testdata/logic_test/txn, line 58 at r1](https://beta.reviewable.io/reviews/cockroachdb/cockroach/23816#-L9lGqhu0EF2n_05-cmZ:-L9lGqhu0EF2n_05-cm_:b-3nrtdt) ([raw file](https://github.com/cockroachdb/cockroach/blob/0fe903cf72156a3074c1b47aaa40ee1273ca7ce7/pkg/sql/logictest/testdata/logic_test/txn#L58)):*
> ```
> 
> statement ok
> create table m3 (id integer, acol varchar(20), bcol timestamp)
> ```

why this random schema with an unused col and why the funky string value below?

---


*Comments from [Reviewable](https://beta.reviewable.io/reviews/cockroachdb/cockroach/23816#-:-L9lH2KLBZ4Sr355fbmS:b-cfofdg)*
<!-- Sent from Reviewable.io -->

andreimatei commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Review status: 0 of 2 files reviewed at latest revision, 3 unresolved discussions, some commit checks failed.

---



*Comments from [Reviewable](https://beta.reviewable.io/reviews/cockroachdb/cockroach/23816#-:-L9lHGWPEdp7b25grIXM:bnfp4nl)*
<!-- Sent from Reviewable.io -->

vivekmenezes commented:
Added a unittest to the PR to check that the descriptor does indeed get released when the trasaction receives a retry error

TFTR!

---

Review status: 0 of 3 files reviewed at latest revision, 3 unresolved discussions.

---

*[pkg/sql/logictest/testdata/logic_test/txn, line 14 at r1](https://beta.reviewable.io/reviews/cockroachdb/cockroach/23816#-L9lG6ze7n1pgtr8QDD0:-L9zU5ztAfOX0jfPG44f:brpmqbq) ([raw file](https://github.com/cockroachdb/cockroach/blob/0fe903cf72156a3074c1b47aaa40ee1273ca7ce7/pkg/sql/logictest/testdata/logic_test/txn#L14)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

what's going on here?
</blockquote></details>

removed!

---

*[pkg/sql/logictest/testdata/logic_test/txn, line 49 at r1](https://beta.reviewable.io/reviews/cockroachdb/cockroach/23816#-L9lG_Q-2Fc5hVMw6uQY:-L9zUNWX-Is9L1m8g4HY:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/0fe903cf72156a3074c1b47aaa40ee1273ca7ce7/pkg/sql/logictest/testdata/logic_test/txn#L49)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

Please explain what we're testing. Also I'd move this to the prepare tests. Also consider using a subtest (we support those in logic tests now).
</blockquote></details>

Done.

---

*[pkg/sql/logictest/testdata/logic_test/txn, line 58 at r1](https://beta.reviewable.io/reviews/cockroachdb/cockroach/23816#-L9lGqhu0EF2n_05-cmZ:-L9zW5yYDJkWSoBd-kZO:bxjvbo8) ([raw file](https://github.com/cockroachdb/cockroach/blob/0fe903cf72156a3074c1b47aaa40ee1273ca7ce7/pkg/sql/logictest/testdata/logic_test/txn#L58)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

why this random schema with an unused col and why the funky string value below?
</blockquote></details>

cleaned up

---


*Comments from [Reviewable](https://beta.reviewable.io/reviews/cockroachdb/cockroach/23816#-:-L9zW83g-IpQolovwitp:bfz4fp8)*
<!-- Sent from Reviewable.io -->

andreimatei commented:
Review status: 0 of 3 files reviewed at latest revision, 3 unresolved discussions.

---

*[pkg/sql/lease_test.go, line 776 at r2](https://beta.reviewable.io/reviews/cockroachdb/cockroach/23816#-LAEa-zeEoCL4wurK7Ay:-LAEa-zeEoCL4wurK7Az:b-rtqvl1) ([raw file](https://github.com/cockroachdb/cockroach/blob/c80b9bd4efa70613eb9722be3b14fe3becf4cd3d/pkg/sql/lease_test.go#L776)):*
> ```Go
> 		SQLLeaseManager: &sql.LeaseManagerTestingKnobs{
> 			LeaseStoreTestingKnobs: sql.LeaseStoreTestingKnobs{
> 				RemoveOnceDereferenced: true,
> ```

a comment on why `RemoveOnceDereferenced` would help.
Sounds like a hack, right? We want to test something about `tableCollection`, but we're using this `RemoveOnceDereferenced` to cause a side-effect that can be observed? Isn't there some reference counter somewhere we could test instead?

---

*[pkg/sql/lease_test.go, line 836 at r2](https://beta.reviewable.io/reviews/cockroachdb/cockroach/23816#-LAEaFdp1mTprjVtXRB_:-LAEaFdp1mTprjVtXRBa:blfpr50) ([raw file](https://github.com/cockroachdb/cockroach/blob/c80b9bd4efa70613eb9722be3b14fe3becf4cd3d/pkg/sql/lease_test.go#L836)):*
> ```Go
> 	}
> 
> 	// Now wait for the release to happen. We use a local timer
> ```

why not use the standard `SucceedsSoon` ?

---


*Comments from [Reviewable](https://beta.reviewable.io/reviews/cockroachdb/cockroach/23816)*
<!-- Sent from Reviewable.io -->

vivekmenezes commented:
Review status: 0 of 3 files reviewed at latest revision, 5 unresolved discussions.

---

*[pkg/sql/lease_test.go, line 776 at r2](https://beta.reviewable.io/reviews/cockroachdb/cockroach/23816#-LAEa-zeEoCL4wurK7Ay:-LAEw70mBcGgS8btmM1O:bnij7z5) ([raw file](https://github.com/cockroachdb/cockroach/blob/c80b9bd4efa70613eb9722be3b14fe3becf4cd3d/pkg/sql/lease_test.go#L776)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

a comment on why `RemoveOnceDereferenced` would help.
Sounds like a hack, right? We want to test something about `tableCollection`, but we're using this `RemoveOnceDereferenced` to cause a side-effect that can be observed? Isn't there some reference counter somewhere we could test instead?
</blockquote></details>

Yeah it's a bit of a hack. There is no counter at the moment. I've added a comment to this.

---

*[pkg/sql/lease_test.go, line 836 at r2](https://beta.reviewable.io/reviews/cockroachdb/cockroach/23816#-LAEaFdp1mTprjVtXRB_:-LAEvqxY1z55elIS2Cpe:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/c80b9bd4efa70613eb9722be3b14fe3becf4cd3d/pkg/sql/lease_test.go#L836)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

why not use the standard `SucceedsSoon` ?
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://beta.reviewable.io/reviews/cockroachdb/cockroach/23816)*
<!-- Sent from Reviewable.io -->

vivekmenezes commented:
bors r+

andreimatei commented on pkg/sql/logictest/testdata/logic_test/prepare:
```diff
@@ -416,3 +416,19 @@ PREPARE x21701c AS SELECT * FROM kv WHERE k IS NOT DISTINCT FROM $1
+PREPARE forbar AS insert into bar (id) VALUES (1)
+
+statement ok
+COMMIT TRANSACTION
```
weird char
